### PR TITLE
Adjust needs system with varying values and external configuration

### PR DESCRIPTION
**Pull request by AndyTechGuy**
Adds a few improvements to the AI character needs system:

- Removes the switch statements used in the old character needs system to a needs list, which allows for easy addition of new kinds of needs.
- Need variances are now created between different characters. This way all characters will get thirsty/hungry at a different rate than every other character, increasing character diversity.
- Allows for external configuration of the needs system. A default config prefab is provided, but can be replaced by a configuration in a different module that has a higher priority.

### CODE REVIEW

File: src/main/java/org/terasology/metalrenegades/ai/component/NeedsComponent.java [hunk 1]
```diff
@@ -15,11 +15,16 @@
...
+import com.google.common.collect.Lists;
 import org.terasology.entitySystem.Component;
 import org.terasology.metalrenegades.ai.CitizenNeed;
 import org.terasology.network.Replicate;
 
+import java.util.Collection;
 import java.util.EnumMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
```
Comment (mayant15) [inline src/main/java/org/terasology/metalrenegades/ai/component/NeedsComponent.java]: Unused imports

File: src/main/java/org/terasology/metalrenegades/ai/component/NeedsComponent.java [hunk 2]
```diff
@@ -28,23 +33,7 @@
...
-
-    public NeedsComponent(CitizenNeed hunger, CitizenNeed thirst, CitizenNeed social, CitizenNeed rest) {
-        this.hungerNeed = hunger;
-        this.thirstNeed = thirst;
-        this.socialNeed = social;
-        this.restNeed = rest;
-    }
+    public List<CitizenNeed> needs = Lists.newArrayList();;
 
     public NeedsComponent() {
```
Comment (mayant15) [inline src/main/java/org/terasology/metalrenegades/ai/component/NeedsComponent.java]: If this constructor does nothing then it's probably cleaner to remove it and use Java's default

File: src/main/java/org/terasology/metalrenegades/ai/system/NeedsSystem.java [hunk 3]
```diff
@@ -18,35 +18,94 @@
...
+            if (configComponent == null || configComponent.priority < newConfig.priority) {
+                configComponent = newConfig;
+            }
+        }
+    }
+
+    @ReceiveEvent
+    public void onCitizenSpawned(CitizenSpawnedEvent citizenSpawnedEvent, EntityRef citizen) {
+        NeedsComponent needsComponent = new NeedsComponent();
+        configComponent.needsConfigs.stream().forEach(config -> {
```
Comment (mayant15) [inline src/main/java/org/terasology/metalrenegades/ai/system/NeedsSystem.java]: [Lists](https://docs.oracle.com/javase/8/docs/api/java/util/List.html) (iterables) don't need a stream and can directly [`forEach`](https://docs.oracle.com/javase/8/docs/api/java/lang/Iterable.html#forEach-java.util.function.Consumer-)

**Review by mayant15 — CHANGES_REQUESTED** [review]

File: src/main/java/org/terasology/metalrenegades/ai/component/NeedsComponent.java (hunk 2 above)
Comment (AndyTechGuy) [reply src/main/java/org/terasology/metalrenegades/ai/component/NeedsComponent.java]: Agreed, the extra constructor has been removed 👍

**Review by mayant15 — APPROVED** [review]

**Pull request merged**